cparser/Elab.ml

```diff
@@ -2383,32 +2446,34 @@ let rec elab_stmt env ctx s =
       | TVoid _, None -> ()
       | TVoid _, Some _ ->
           error loc
-            "'return' with a value in a function of return type 'void'"
+            "'return' with a value in a function returning void"
       | _, None ->
-          warning loc
-            "'return' without a value in a function of return type@ %a"
-            Cprint.typ ctx.ctx_return_typ
+          warning loc Return_type
+            "'return' with no value, in a function returning non-void"
       | _, Some b ->
           if not (wrap2 valid_assignment loc env b ctx.ctx_return_typ)
           then begin
             if wrap2 valid_cast loc env b.etyp ctx.ctx_return_typ then
-              warning loc
-                "return value has type@ %a@ \
-                 instead of the expected type@ %a"
-                Cprint.typ b.etyp Cprint.typ ctx.ctx_return_typ
+              if wrap2 int_pointer_conversion loc env b.etyp ctx.ctx_return_typ then
+                warning loc Int_conversion
+                  "incompatible integer to pointer conversion returning %a from a function with result type %a"
+                  (print_typ env) b.etyp (print_typ env) ctx.ctx_return_typ
+              else
+                warning loc Int_conversion
+                  "incompatible integer to pointer conversion returning %a from a function with result type %a"
+                  (print_typ env) b.etyp (print_typ env) ctx.ctx_return_typ
```

Comment:
This should like the one above be int => pointer in the if branch and pointer to int in the else branch.